test/functional/ui/cursor_spec.lua

```diff
@@ -18,138 +18,155 @@ describe('ui/cursor', function()
   end)
 
   it("'guicursor' is published as a UI event", function()
-    local expected_cursor_style = {
-      cmdline_hover = {
-        mouse_shape = 0,
-        short_name = 'e' },
-      cmdline_insert = {
+    local expected_mode_info = {
+      [1] = {
         blinkoff = 250,
         blinkon = 400,
         blinkwait = 700,
-        cell_percentage = 25,
-        cursor_shape = 'vertical',
+        cell_percentage = 0,
+        cursor_shape = 'block',
+        full_name = 'normal',
         hl_id = 46,
         id_lm = 47,
         mouse_shape = 0,
-        short_name = 'ci' },
-      cmdline_normal = {
+        short_name = 'n' },
+      [2] = {
         blinkoff = 250,
         blinkon = 400,
         blinkwait = 700,
         cell_percentage = 0,
         cursor_shape = 'block',
+        full_name = 'visual',
         hl_id = 46,
         id_lm = 47,
         mouse_shape = 0,
-        short_name = 'c' },
-      cmdline_replace = {
+        short_name = 'v' },
+      [3] = {
         blinkoff = 250,
         blinkon = 400,
         blinkwait = 700,
-        cell_percentage = 20,
-        cursor_shape = 'horizontal',
+        cell_percentage = 25,
+        cursor_shape = 'vertical',
+        full_name = 'insert',
         hl_id = 46,
         id_lm = 47,
         mouse_shape = 0,
-        short_name = 'cr' },
-      insert = {
+        short_name = 'i' },
+      [4] = {
         blinkoff = 250,
         blinkon = 400,
         blinkwait = 700,
-        cell_percentage = 25,
-        cursor_shape = 'vertical',
+        cell_percentage = 20,
+        cursor_shape = 'horizontal',
+        full_name = 'replace',
         hl_id = 46,
         id_lm = 47,
         mouse_shape = 0,
-        short_name = 'i' },
-      more = {
-        mouse_shape = 0,
-        short_name = 'm' },
-      more_lastline = {
-        mouse_shape = 0,
-        short_name = 'ml' },
-      normal = {
+        short_name = 'r' },
+      [5] = {
         blinkoff = 250,
         blinkon = 400,
         blinkwait = 700,
         cell_percentage = 0,
         cursor_shape = 'block',
+        full_name = 'cmdline_normal',
```

Comment:
I like to turn that question around, what so special about `full_name` it deserves to be the key? For the TUI (and other clients only caring about style) before this change there was a needless transformation from an array to a dict and back to an array again... And for clients that want to interpret the mode, it would be better with fields that actually describe the mode (mapping mode, selection type, temporary substate). I will add that in a later PR. This string should be mostly useful for testing and debugging purposes.